File: stable/tom/library/sl/MuVar.java
package tom.library.sl;

/**
 * <p>
 * Basic strategy combinator with one argument <code>name</code> 
 * which denotes the name of the mu-variable
 * This basic visitor combinator used to build recursive visitors
 * After the mu-expansion, the private variable <code>instance</code>
 * is a pointer to the strategy to execute
 * <p>
 */

public class MuVar extends AbstractStrategyCombinator {
  private Strategy instance = null;
  protected String name;
  
  public MuVar(String name) {
    // make sure the MuVar is seen as a leaf for all visitors
    initSubterm();
    this.name = name;
  }

  public boolean equals(Object o) {
    if(o instanceof MuVar) {
      MuVar muVar = (MuVar)o;
      if(muVar != null) {
        if(name != null) {
          return name.equals(muVar.name);
        } else {
          return name==muVar.name && instance==muVar.instance;
        }
      }
    }
    return false;
  }

  public int hashCode() {
    if(name!=null) {
      return name.hashCode();
    } else {
      return instance.hashCode();
    }
  }
  
  public final <T> T visitLight(T any,Introspector i) throws VisitFailure {
    if(instance != null) {
      return instance.visitLight(any,i);
    } else {
      throw new VisitFailure();
    }
  }

  public int visit(Introspector i) {
    if(instance != null) {
      return instance.visit(i);
    } else {
      //setStatus(Environment.FAILURE);
      return Environment.FAILURE;
    }
  }

  public Strategy getInstance() {
    return instance;
  }
  
  public void setInstance(Strategy v) {
    this.instance = v;
  }
  
  public void setName(String name) {
    this.name = name;
  }

  public final boolean isExpanded() {
    return instance != null;
  }

  public String getName() {
    return name;
  }

  public String toString() {
    return "[" + name + "," + instance + "]";
  }
}

File: tomgo/stable/library/sl/mu_var.go
package sl

// MuVar is the named-variable side of the [Mu] fixed-point combinator.
// Until [Mu] expansion, MuVar is an "unbound" leaf that fails on any
// VisitLight. Expansion sets Instance to the recursive body, after
// which VisitLight delegates to it.
//
// MuVar instances are matched by name: two MuVars are considered the
// same variable if their names match, regardless of identity. This
// lets users write a strategy graph with multiple MuVar("x") leaves
// and have all of them bind to the same Mu body.
//
// Equivalent to Java's tom.library.sl.MuVar.
type MuVar struct {
	abstractCombinator
	Name     string
	Instance Strategy // nil until parent Mu is expanded
}

func NewMuVar(name string) *MuVar {
	m := &MuVar{Name: name}
	m.initSubterm()
	return m
}

func (m *MuVar) VisitLight(subject any, intro Introspector) (any, error) {
	if m.Instance == nil {
		return nil, ErrVisitFailure
	}
	return m.Instance.VisitLight(subject, intro)
}

// IsExpanded reports whether this MuVar has been bound to a Mu body
// by [Mu.expand].
func (m *MuVar) IsExpanded() bool { return m.Instance != nil }
